lib/types.ts

```diff
@@ -0,0 +1,85 @@
+import type * as typesObject from './constant/types'
+import type {ValueOf} from 'type-fest'
+
+export interface Point {
+  line: number
+  column: number
+  offset: number
+  _index: number
+  _bufferIndex: number
+}
+export interface Token {
+  start: Point
+  end: Point
+  type: unknown
+  contentType: string
+  previous: Token
+  next: Token
+  _size: number
+  _break?: boolean
+  _tokenizer?: unknown
+  _events?: Event[]
+  _subevents?: Event[]
+  _contentTokenized?: unknown
+  _close?: unknown
+  _marker?: unknown
+  _open?: unknown
+  _side?: number
+  _tabs?: unknown
+}
+
+export type Type = ValueOf<typeof typesObject>
+
+export type Event = [name: string, token: Token, context: unknown]
+
+export interface Effects {
+  enter: (type: Type) => Token
+  consume: (code: number) => Token
+  exit: (type: Type) => Token
+  attempt: (
+    one: unknown,
+    two: unknown,
+    three?: unknown
+  ) => (code: number) => unknown
```

Comment:
Attempt deals with several values, and tries to parse according to those values. If a value resulted in `ok`, it worked, the tokens that were made are used, and `returnState` (two) is switched to. If the result is `nok`, the attempt failed, so we revert to the original state, and `bogusState` (three) is used.

The values that can be given are:
- An object of “hooks”, where the keys are character codes and the values are either a single construct or an array of constructs. If the first character that is attempted matches, the corresponding constructs are attempted (see next), otherwise: bogusState
- A list of constructs, which will be attempted in order. If one results in `nok`, we revert and the next one is attempted
- A single constructs (an object with a `tokenize` function, and potentially some other stuff)

Reply:
Added some follow up questions in https://github.com/micromark/micromark/pull/17#discussion_r486744591  and https://github.com/micromark/micromark/pull/17#discussion_r486744731